ui/src/main/java/org/jtaccuino/core/ui/completion/CompletionSelectionModel.java

```diff
@@ -0,0 +1,224 @@
+            while (c.next()) {
+                if (c.wasAdded() || c.wasRemoved() || c.wasPermutated()) {
+                    int currentSelected = getSelectedIndex();
+                    if (currentSelected != -1 && (currentSelected >= items.size() || items.get(currentSelected) == null)) {
+                        int newSelection = findNextSelectableIndex(currentSelected, false); // Try forward from current index
+                        if (newSelection == -1) {
+                            newSelection = findPrevSelectableIndex(currentSelected - 1, false);
+                        }
+                        if (newSelection != -1) {
+                            select(newSelection);
+                        } else {
+                            clearSelection();
+                        }
+                    }
+                }
+            }
+        });
+    } // Try forward from current index
+
+    @Override
+    public ObservableList<Integer> getSelectedIndices() {
+        return selectedIndicesList;
+    }
+
+    @Override
+    public ObservableList<CompletionItem> getSelectedItems() {
+        return selectedItemsList;
+    }
+
+    @Override
+    public void select(int index) {
+        clearAndSelect(index);
+    }
+
+    @Override
+    public void select(CompletionItem obj) {
+        int index = items.indexOf(obj);
+        if (index != -1 && obj != null) {
+            select(index);
+        } else {
+            clearSelection();
+        }
+    }
+
+    @Override
+    public void selectNext() {
+        int currentIndex = getSelectedIndex();
+        int nextIndex = findNextSelectableIndex(currentIndex + 1, false);
+        select(nextIndex);
+    }
+
+    @Override
+    public void selectPrevious() {
+        int currentIndex = getSelectedIndex();
+        int prevIndex = findPrevSelectableIndex(currentIndex - 1, false);
+        select(prevIndex);
+    }
+
+    private int findNextSelectableIndex(int startIndex, boolean wrapAround) {
+        int maxIndex = items.size();
+        for (int i = startIndex; i < maxIndex; i++) {
+            if (i >= 0 && i < items.size()) {
+                org.jtaccuino.core.ui.completion.CompletionItem item_at_i = items.get(i);
+                if (!CompletionItem.NIL.equals(item_at_i)) {
+                    return i;
+                }
+            }
+        }
+        if (wrapAround) {
+            for (int i = 0; i < startIndex; i++) {
+                if (i >= 0 && i < items.size()) {
+                    org.jtaccuino.core.ui.completion.CompletionItem item_at_i = items.get(i);
+                    if (!CompletionItem.NIL.equals(item_at_i)) {
+                        return i;
+                    }
+                }
+            }
+        }
+        return -1;
+    }
+
+    private int findPrevSelectableIndex(int startIndex, boolean wrapAround) {
+        int minIndex = 0;
+        for (int i = startIndex; i >= minIndex; i--) {
```

Comment:
variable declared and used only once

```suggestion
        for (int i = startIndex; i >= 0; i--) {
```